---
title: "tutorial_DA"
author: "jjz"
date: "2020/9/15"
output: html_document
---

## Tutorial on Data analysis

***

### Step 1
##### Upload the metadata (.tsv) and taxonomic table file (.qza)
1. The first column name of the metadata must be **#SampleID**.
2. The taxonomic table file can be downloaded from **Sequences preprocessing/Taxonomic analysis**.
3. If you have a taxonomic table file (.qza) that you yourself derived from QIIME2, it may be uploaded at this stage.
4. After you finish uploading, the results will be presented.

<img src="/home/imuser/text_files/p4_1.png" width="70%" />

##### You can download the example files to inspect the results of this process.
1. Download **Metadata_example.tsv** and upload it.
2. Download **Taxonomic_table_example.qza** and upload it.

<img src="/home/imuser/text_files/p4_2.png" width="40%" />

3. After uploading, the 8 results will be presented.

<img src="/home/imuser/text_files/p4_3.png" width="100%" />

***

### Step 2
##### Inspect the 8 results.
##### 1. **Taxonomic table:**A table in which the first 7 columns contain taxonomy information and the other columns contain values of read counts.

+ **Choose the group:** Select the group. The read count table will then be grouped accordingly, based on the metadata.
+ **Taxonomy information:** In this table, the first 7 columns contain taxonomy information. K denotes the number of taxa at a given level.
+ **Read count:** In this region, the column names are the sample IDs and the values are read counts.
+ **Download table**: Click on the button **"Download Taxonomic table"** to download the table.

<img src="/home/imuser/text_files/p4_4.png" width="100%" />

--

##### 2. **Taxonomic barplot:** An interactive barplot which shows the percentages of all species in a sample. Each categoric variable is represented by a bar.

+ **Interactive barplot:** When the cursor is on the bar region, the species information will be presented. Click and drag on the plot to zoom in and out, and double click on the plot to convert the information to its original state.
+ **Choose the level:** Select a level. The taxa in the plot will then be presented at that level.
+ **Top N:** When you select N, the plot will show the union of the N taxa with the highest relative abundance in all samples. For example, the top 2 taxa in  Sample A is taxa_1 and taxa_2. The top 2 species in Sample B is taxa_1 and taxa_3. Consequently, the plot will show the relative abundance of taxa_1, taxa_2 and taxa_3 in all samples.
+ **Choose the group:** Select the group. The barplot will then be grouped based on the metadata.
+ **Download barplot**: Click on the button **"Download barplot"** to download the barplot. Alternatively, click on the camera icon in the plot.
    
<img src="/home/imuser/text_files/p4_5.png" width="100%" />

<img src="/home/imuser/text_files/p4_6.png" width="100%" />

--

##### 3. **Taxonomic heatmap:** An interactive heatmap which shows the log10-transformation percentages of all species in a sample. All values are automatically increased by 0.01 in order to prevent zeros. The values are shown in various colors.

+ **Interactive heatmap:** When the cursor is on the bar, the species information will be presented. Click and drag on the plot to zoom in and out, and double click on the plot to convert the information to its original state. You can download the plot by clicking on the camera icon in the plot.
+ **Choose the level:** Select a level. The taxa in the plot will then be presented at that level.
+ **Choose the group:** Select the group. The heatmap will then be grouped based on the metadata accordingly.
+ **Download heatmap matrix**: Click on the button **"Download Heatmap matrix"** to download the heatmap matrix data.
    
<img src="/home/imuser/text_files/p4_7.png" width="100%" />

<img src="/home/imuser/text_files/p4_8.png" width="100%" />

--

##### 4. **Krona:** A visualization tool which allows hierarchical data to be explored with zooming, multi-layered pie charts. [Get more information](https://github.com/marbl/Krona/wiki)

+ **Krona pie chart:** An interactive pie plot. Double click on a selection of taxa at a given level to inspect the taxa at the sublevel.
+ **Choose the sample:** Select the sample to switch to the corresponding pie plot.
+ **Download Krona results:** Click on the button **"Download Krona results (.zip)"** to download the interactive pie plot (html files with Javascript).

<img src="/home/imuser/text_files/p4_9.png" width="100%" />

--

##### 5. **Alpha diversity:** Evaluation of species diversity in a sample. In MOCHI, we adapt 8 indexes (richness, Chao1, ACE, Shannon diversity, InvSimpson diversity, Shannon evenness, Simpson evenness, Goods coverage).

* **Table**
    + **Alpha diversity table:** This table shows the values of 8 alpha diversity indexes.
    + **Download Alpha diversity table:** Click on the button **"Download Alpha Diversity Table"** to download the table.

<img src="/home/imuser/text_files/p4_10.png" width="100%" />

* **Boxplot**
    + **Choose the index:** Select an index from the drop-down menu. A boxplot will then be presented with that index.
    + **Choose the statistic method:** Select ANOVA (parametric method) or Kruskal-Wallis (nonparametric method) to test whether the distribution of the index is significantly different among the groups.
    + **Choose the group:** Select the group. The values of the index in the boxplot will then be grouped accordingly, based on the metadata.
    + **Download Alpha diversity boxplot:** Click on the button **"Download Alpha Diversity Boxplot"** to download the boxplot.

<img src="/home/imuser/text_files/p4_11.png" width="100%" />

* **Post-hoc analysis**
    + If ANOVA is selected when creating the boxplot, then the Tukey test will be used for the post-hoc test; if Kruskal-Wallis is selected, then the Dunn test will be used.
    + **Download Alpha diversity post-hoc test result:** Click on the button **"Download Alpha Diversity statistical result"** to download the post-hoc test result.

<img src="/home/imuser/text_files/p4_12.png" width="100%" />

--

##### 6. **Beta diversity:** Evaluation of species diversity among samples. In MOCHI, we use the Bray-Curtis index.

* **Distance matrix**
    + **Interactive heatmap:** When the cursor is on the heatmap, the distance will be presented. Click and drag on the plot to zoom in and out, and double click on the plot to convert the information to its original state. You can download the plot by clicking on the camera icon in the plot.
    + **Download distance matrix:** Click on the button **"Download Beta Diversity distance matrix data"** to download the matrix data.

<img src="/home/imuser/text_files/p4_13.png" width="100%" />

* **Dimension-reduction plot**
    + We provide three dimension-reduction methods, including PCA (Principal Component Analysis), PCoA (Principal Co-ordinates Analysis) and NMDS (Non-metric Multidimensional Scaling) for visualization of beta diversity.
    + **Choose the method:** Select the dimension reduction methods to be used.
    + **Clustering with circle:** If you check this checkbox, the samples in the same group will be grouped with a circle in the plot.
    + **Choose the group:** The samples in the plot will be labeled using colors based on the metadata.
    + **Download beta diversity plot:** Click on the button **"Download Beta Diversity Plot"** to download the plot.

<img src="/home/imuser/text_files/p4_14.png" width="100%" />

* **Statistical analysis**
    + We provide three statistical methods, including PerMANOVA (Permutational Multivariate Analysis of Variance), ANOSIM (Analysis of Similarities) and MRPP (Multiple Response Permutation Procedure), to test whether beta diversity is significantly different among groups or between pairs of groups.
    + **Download statistical result table:** Click on the button below the table to download the statistical results.

<img src="/home/imuser/text_files/p4_15.png" width="100%" />

--

##### 7. **Phylogenetic diversity:** A species diversity which considers the genetic distance. In MOCHI, we use Faith PD (a kind of alpha diversity which considers the genetic distance) and Unifrac distance (a kind of beta diversity whcih considers the genetic distance).

* **Upload the files:** Upload the feature table (.qza) and the sequences file (.qza). If you have already finished the **sequences preprocessing**, you need only  download two files from **Sequences preprocessing/Taxonomic analysis** and upload them. Also, if your feature table file and sequences file (.qza) were derived from QIIME2 by you, these files can be uploaded at this stage.
* **Sampling depth:** If the total count for any sample(s) is samller than this value, those samples will be dropped from the diversity analysis. The default value is the total count of a sample with the smallest count. This means that no sample with default value will be dropped.
* **Number of threads:** The number of threads to use for multithreaded process. The default value is all threads-2.
* **Start:** Click on the button **"Start!"** to execute the analysis after the above files and parameters have been prepared.

<img src="/home/imuser/text_files/p4_16.png" width="100%" />

* **Faith PD table:** Faith PD (Faith's Phylogenetic Diversity) is a commonly used phylogenetic index. PD is a species diversity that considers genetic distance among species.
    + **Faith PD table:** This table shows the Faith PD (phylogenetic diversity) of all samples.
    + **Download Faith PD table:** Click on the button **"Download Faith PD table"** to download the table.
    
<img src="/home/imuser/text_files/p4_17.png" width="100%" />

* **Faith PD boxplot:** This shows the distribution of Faith PD using a boxplot.
    + **Choose the statistical method:**  Select ANOVA (parametric method) or Kruskal-Wallis (nonparametric method) to test whether the distribution of Faith PD is significantly different among the groups.
    + **Choose the group:** Select a group. Faith PD will then be grouped accordingly in the boxplot, based on the metadata.
    + **Download Faith PD boxplot:**  Click on the button **"Download Faith PD Boxplot"** to download the boxplot.

<img src="/home/imuser/text_files/p4_18.png" width="100%" />

* **Post-hoc analysis** 
    + If ANOVA is selected in the boxplot step, the Tukey test will be used for post-hoc test; if Kruskal-Wallis is selected, then the Dunn test will be used.
    + **Download Faith PD post hoc result:** Click on the button **"Download Faith PD post hoc result"** to download the result.
    
<img src="/home/imuser/text_files/p4_19.png" width="50%" />

* **Heatmap of UniFrac distance** 
    + **Interactive heatmap:** When the cursor is on the bar, the speceis information will be presented. Click and drag on the plot to zoom in and out, and double click on the plot to convert the information to its original state. You can download the plot by clicking on the camera icon in the plot.
    + **Choose the method:** Select unweighted UniFrac (do not consider the richness of taxa) or weighted UniFrac (consider the richness of taxa).
    + **Download heatmap matrix**: Click on the button **"Download UniFrac distance matrix"** to download the matrix data.

<img src="/home/imuser/text_files/p4_20.png" width="100%" />

* **Dimension-reduction plot of UniFrac distance**
    + We provide three dimension-reduction methods, including PCoA (Principal Co-ordinates Analysis) and NMDS (Non-metric Multidimensional Scaling) for visualization of UniFrac distance.
    + **Choose the ordination method:** Select the dimension-reduction methods for to be used for creating the plot.
    + **Clustering with circle:** If you check this checkbox, the samples in the same group will be grouped with a circle in the plot.
    + **Choose the group:** The samples in the plot will be labeled using colors based on the metadata.
    + **Download beta diversity plot:** Click on the button **"Download Beta Diversity Plot"** to download the plot.

<img src="/home/imuser/text_files/p4_21.png" width="100%" />

* **Statistical analysis**
    + We provide three statistical methods, including PerMANOVA (Permutational Multivariate Analysis of Variance), ANOSIM (Analysis of Similarities) and MRPP (Multiple Response Permutation Procedure), to test whether beta diversity is significantly different among groups or between pairs of groups.
    + **Download statistical result table:** Click on the button below the table to download the statistical results.

<img src="/home/imuser/text_files/p4_22.png" width="100%" />

--

##### 8. **ANCOM:** Analysis of composition of microbiomes. This is used for comparing the composition of microbiomes in two or more populations. [Get more information](https://www.tandfonline.com/doi/full/10.3402/mehd.v26.27663)

* **Choose the target comparison:** Select a group. ANCOM will then find significantly different abundant taxa among subgroups in that group.
* **Start:** Click on the button **"Start!"** to execute the analysis after the target comparison is determined.

<img src="/home/imuser/text_files/p4_23.png" width="100%" />

* **ANCOM volcano plot:** This interactive plot shows the significantly different abundant taxa (red). When the cursor is on the dot, the taxa information will be presented.

<img src="/home/imuser/text_files/p4_24.png" width="100%" />

* **ANCOM statistical results:** This table shows the W values for all taxa.
    + **Download the ANCOM result table:** Click on the button **"Download the ANCOM result table"** to download the results. The table will contain the W values for  all taxa.

<img src="/home/imuser/text_files/p4_25.png" width="100%" />